### ISSUE TITLE
model yield residuals

### ISSUE THREAD

**Issue by krsnapaudel** [BigDataWUR/AgML-CY-Bench#277]
Add scikit-learn and lstm models that leverage the trend model (or average yield model) to remove the trend (or mean) and train the ML/DL models to predict the residuals from the trend (or mean).

The predictions or forecasts will again use the trend model (or average yield model) to predict the trend (or mean) and add that back to the prediction of ML/DL models.

NOTE: These models must be separate from those that predict yield.

Also see the discussion here: https://github.com/BigDataWUR/AgML-CY-Bench/discussions/292

**Comment by krsnapaudel** [comment]
Our dataset API is not friendly to modify labels and model residuals. Example for Ridge.

```
class RidgeRes(SklearnRidge):
    def __init__(self, feature_cols: list = None):
        self._trend_model = TrendModel()

    def fit(self, dataset: Dataset):
        train_df = data_to_pandas(dataset, data_cols=[KEY_LOC, KEY_YEAR, KEY_TARGET])
        train_residuals = []
        for item in dataset:
            trend_train_df = train_df[(train_df[KEY_LOC] == item[KEY_LOC]) &
                                      (train_df[KEY_YEAR] != item[KEY_YEAR])]
            # only one year of training data for this location
            if (trend_train_df.empty):
                trend_pred = train_df[KEY_TARGET].mean()
            else:
                trend_train_df.set_index([KEY_LOC, KEY_YEAR], inplace=True)
                trend_train_dataset = Dataset(dataset.crop, data_target=trend_train_df)
                self._trend_model.fit(trend_train_dataset)
                trend_pred, _ = self._trend_model.predict_items([item])

            res_row = {
                KEY_LOC: item[KEY_LOC],
                KEY_YEAR: item[KEY_YEAR],
                KEY_TARGET: trend_pred - item[KEY_TARGET]
            }        
            train_residuals.append(res_row)

        residuals_df = pd.DataFrame(train_residuals)
        residuals_df.set_index([KEY_LOC, KEY_YEAR], inplace=True)
        # Now we need another instance of dataset with residuals_df as data_targets.
```
How do we do this without duplicating data? Should we just duplicate data?
Any thoughts, @aikepotze, @michielkallenberg?

**Comment by aikepotze** [comment]
I think the easiest solution is to allow the fit method of the BaseSklearnModel to accept an optional target dataframe, that overrides the target values if provided and is indexed by the same keys.

Alternatively we could create a dataset wrapper. But that might be more complicated.
